app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,193 @@
+      const deref = weakReffedValue[weakReffedValueKey].deref();
+      if (deref === undefined) {
+        this.map.delete(weakReffedValueKey);
+        return undefined;
+      }
+      dereffedValue[weakReffedValueKey] = deref;
+    }
+
+    return dereffedValue;
+  }
+
+  /**
+   * Checks whether the map contains the specified key.
+   * @param key The key to check for presence in the map.
+   * @returns `true` if the map contains the key, otherwise `false`.
+   */
+  has(key: string): boolean {
+    return this.get(key) !== undefined;
+  }
+
+  /**
+   * Removes the specified key and its associated value from the map.
+   * @param key The key to remove along with its associated value.
+   * @returns `true` if the element was successfully removed, otherwise `false`.
+   */
+  delete(key: string): boolean {
+    const value = this.get(key);
+    if (value !== undefined) {
+      return this.map.delete(key);
+    }
+    return false;
+  }
+
+  /**
+   * Removes all key-value pairs from the map.
+   */
+  clear() {
+    this.map.clear();
+  }
+
+  /**
+   * Returns the number of key-value pairs present in the map.
+   */
+  get size(): number {
+    return this.map.size;
+  }
+
+  /**
+   * Returns a new iterator object that contains an array of `[key, value]` for each element in the map.
+   * The values are dereferenced before being returned.
+   */
+  entries(): IterableIterator<[string, object]> {
+    const entries: Array<[string, object]> = [];
+    this.map.forEach((value, key) => {
+      const derefValue = this.get(key);
+      if (derefValue !== undefined) {
+        entries.push([key, derefValue]);
+      }
+    });
+    return entries.values();
+  }
+
+  /**
+   * Returns a new iterator object that contains the keys for each element in the map.
+   */
+  keys(): IterableIterator<string> {
+    return this.map.keys();
+  }
+
+  /**
+   * Returns a new iterator object that contains the values for each element in the map.
+   * The values are dereferenced before being returned.
+   */
+  values(): IterableIterator<object> {
+    const values: Array<object> = [];
+    this.map.forEach((_, key) => {
+      const derefValue = this.get(key);
+      if (derefValue !== undefined) {
+        values.push(derefValue);
+      }
+    });
+    return values.values();
+  }
+
+  /**
+   * Returns a new iterator object that contains an array of `[key, value]` for each element in the map,
+   * making the map itself iterable.
+   */
+  [Symbol.iterator](): IterableIterator<[string, object]> {
+    return this.entries();
+  }
+
+  /**
+   * Returns a string representing the map. This is used when converting the map to a string,
+   * e.g., by `Object.prototype.toString`.
+   */
+  get [Symbol.toStringTag](): string {
+    return 'WeakRefObjectMap';
+  }
+
+  /**
+   * Executes a provided function once for each key-value pair in the map, in insertion order.
+   * Note that the values passed to the callback function are the `WeakRefCompatibleObject`s,
```

Comment:
Hmm; making the type of the values in this callback would conflict with the types for `Map`.

It sounds potentially useful, but for type compatibility we might want to make this a separate method if it's something we need.